pgjdbc/src/main/java/org/postgresql/jdbc/PgBlobBytea.java

```diff
@@ -0,0 +1,238 @@
+                              PSQLState.INVALID_PARAMETER_VALUE);
+    }
+    int res = indexOf(this.data, pattern, (int) start - 1);
+    if (res >= 0) {
+      res = res + 1;
+    }
+    return res;
+  }
+
+  @Override
+  public long position​(Blob pattern, long start) throws SQLException {
+    return position(pattern.getBytes(1, (int) pattern.length()), start);
+  }
+
+  @Override
+  public OutputStream setBinaryStream​(long pos) throws SQLException {
+    return new ByteaBlobOutputStream(pos);
+  }
+
+  @Override
+  public int setBytes​(long pos, byte[] bytes) throws SQLException {
+    return setBytes(pos, bytes, 0, bytes.length);
+  }
+
+  @Override
+  public int setBytes​(long pos, byte[] bytes, int offset, int len) throws SQLException {
+    if (this.data == null) {
+      if (pos == 1) {
+        this.data = new byte[0];
+      } else {
+        throw new PSQLException(GT.tr("Called setBytes on null Blob"),
+                              PSQLState.INVALID_PARAMETER_VALUE);
+      }
+    }
+
+    if (pos < 1 || pos > data.length + 1) {
+      throw new PSQLException(GT.tr("Invalid pos parameter {0} for length {1}", pos, data.length),
+                              PSQLState.INVALID_PARAMETER_VALUE);
+    }
+    if (offset < 0) {
+      throw new PSQLException(GT.tr("Invalid offset parameter {0}", offset),
+                              PSQLState.INVALID_PARAMETER_VALUE);
+    }
+    if (offset + len > bytes.length) {
+      throw new PSQLException(GT.tr("Invalid offset/len parameters {0}/{1} for length {2}", offset, len, bytes.length),
+                              PSQLState.INVALID_PARAMETER_VALUE);
+    }
+    pos--;
+    // extend the array if necessary
+    if (pos + len > data.length) {
+      this.data = Arrays.copyOf(this.data, (int) pos + len);
+    }
+    System.arraycopy(bytes, offset, this.data, (int) pos, len);
+    return len;
+  }
+
+  @Override
+  public void	truncate​(long len) throws SQLException {
+    if (this.data == null) {
+      throw new PSQLException(GT.tr("Called truncate on null Blob}"),
+                              PSQLState.INVALID_PARAMETER_VALUE);
+    }
+    if (len <= data.length) {
+      this.data = Arrays.copyOf(this.data, (int) len);
+    } else {
+      throw new PSQLException(GT.tr("Called truncate with length too long for string of length {0}", data.length),
+                              PSQLState.INVALID_PARAMETER_VALUE);
+    }
+  }
+
+  // shamelessly stolen from
+  // http://www.java2s.com/Code/CSharp/File-Stream/Searchabytearrayforasubbytearray.htm
+  // nice single loop solution
+  private static int indexOf(byte[] array, byte[] pattern, int offset) {
+    int success = 0;
+    for (int i = offset; i < array.length; i++) {
+      if (array[i] == pattern[success]) {
+        success++;
+      } else {
+        success = 0;
+      }
+      if (pattern.length == success) {
+        return i - pattern.length + 1;
+      }
+    }
+    return -1;
+  }
```

Comment:
This doesn't work for repeated values in the pattern. It would need to back track to the start of the match:

```java
byte array[] = {0, 1, 1, 1, 2, 3};

byte found[] = {1, 1, 1, 2};
// indexOf = 1
System.out.println("indexOf = " + indexOf(array, found, 0));

// Subset of previous value should be 2 but it's not found:
byte notFound[] = {1, 1, 2};
// indexOf = -1
System.out.println("indexOf = " + indexOf(array, notFound, 0));
```

Reply:
Ouch. OK, I'll just use a double loop and be done with it. That's what I get for trying to be clever. Thanks for noticing.

Reply:
or maybe I'll use KMP searching :-)